cql3/statements/ks_prop_defs.cc

```diff
@@ -96,14 +96,35 @@ std::map<sstring, sstring> ks_prop_defs::get_replication_options() const {
     return std::map<sstring, sstring>{};
 }
 
+data_dictionary::storage_options ks_prop_defs::get_storage_options() const {
+    data_dictionary::storage_options opts;
+    auto options_map = get_map(KW_STORAGE);
+    if (options_map) {
+        auto set_field = [&options_map] (sstring& var, sstring key) {
+            auto it = options_map->find(key);
+            if (it != options_map->end()) {
+                var = it->second;
+            }
+        };
+        auto it = options_map->find("type");
+        if (it != options_map->end()) {
+            opts.type = data_dictionary::storage_options::parse_type(it->second);
+        }
+        set_field(opts.bucket, "bucket");
+        set_field(opts.key_id, "key_id");
+        set_field(opts.endpoint, "endpoint");
+    }
```

Comment:
let's fail create ks statement if unrecognized option is found

Reply:
Also, the three specific items you copy here (bucket, key_id, endpoint) only makes sense for the S3 type, not for the native part. So why don't we just copy all the options here instead of these specific three, and just the CQL statement to set the options will validate that the combination makes sense, i.e, that with S3 you can specify "bucket" but not "doghouse", and with NATIVE you can't specify neither bucket nor doghouse.

Of course later as we add more types and options, we'll need additional cluster options so you can't set a new type (say "NFS") if not all the nodes on the cluster support it. What a mess :-( But at least the mess will be in one place (the CQL statement validation) and not all over the place as in this function?

Reply:
Back in the hackathon days, it was decided that these "storage options" should simply hold a union of all options needed for different formats that we support. E.g. for a shared mountpoint we can still use the "bucket" part to indicate a local path (which is logically a bucket), and key_id can also make sense for optional authentication. Native storage simply doesn't use any options, which is fine. And if any new storage driver needs another option, it can be added to this structure.

Reply:
If you go with this approach, I think you need to be a bit more forward-looking with the options and their names - e.g., why call something a "bucket" if this name only has meaning for S3? For example, we could call it "location" and then explain in the comment that this is a type-specific string which describes where to store the files - e.g., in s3 it refers to a bucket.

I don't know what key_id was supposed to mean. Is it an authentication key of some sort? There is just one of those (Amazon has two?)?

Anyway, what if I add an NFS storage option and find myself needing a "doghouse" option, but then someone sets type=S3 but doghouse=7? Which part of the code will tell me I did something wrong here?

Reply:
By the way, this discussion reminded me of the idea of URL. It was invented exactly to encapsulate these things - a URL http://some.domain.com/some/directory a "type" (e.g., http://), "endpoint" (some.domain.com/) and directory (some/directory) - and also additional parameters and optional port. 
I'm not saying you should use a URL here, but it's a thought.

Reply:
> something like https://github.com/scylladb/scylla/blob/master/compaction/compaction_descriptor.hh#L52 should do the work. we have to guarantee that each type performs its own validation on its supported options

yeah, I'm already in the middle of implementing this kind of approach. I'll push something today